web/src/util.js

```diff
@@ -1,5 +1,254 @@
-const example = x => x
+import { useState } from "react"
 
-export {
-    example
-}
\ No newline at end of file
+const headerContentTypes = [
+    "application/json",
+    "application/xml",
+    "application/javascript",
+    "application/pdf",
+    "application/zip",
+    "application/octet-stream",
+    "application/ld+json",
+    "application/vnd.api+json",
+    "application/msword",
+    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+    "application/vnd.ms-excel",
+    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+    "application/vnd.ms-powerpoint",
+    "application/vnd.openxmlformats-officedocument.presentationml.presentation",
+    "application/x-www-form-urlencoded",
+    "application/x-tar",
+    "application/x-rar-compressed",
+    "application/x-7z-compressed",
+    "audio/mpeg",
+    "audio/wav",
+    "image/jpeg",
+    "image/png",
+    "image/gif",
+    "image/webp",
+    "image/svg+xml",
+    "multipart/form-data",
+    "text/html",
+    "text/plain",
+    "text/css",
+    "text/javascript",
+    "text/xml",
+    "text/csv",
+    "text/markdown",
+    "video/mp4",
+    "video/mpeg",
+    "video/webm",
+    "video/ogg"
+]
+
+const emptyValue = "<Enter Value>"
+
+const highlightPartOfString = (found, string) => {
+    const index = string.toLowerCase().indexOf(found.toLowerCase())
+    const [prefix, toHighlight, postfix] = [
+        string.slice(0, index),
+        string.slice(index, found.length + index),
+        string.slice(found.length + index, string.length)
+    ]
+    return { prefix, toHighlight, postfix }
+}
+/**
+ *
+ * @param {*} recommendations Options of Autocomplete as an array
+ * @returns
+ */
+const AutoCompleteInput = ({ recommendations }) => {
+    const [inputValue, setInputValue] = useState("")
+    const [firstRecommendation, setFirstRecommendation] = useState("")
+    const [showFirstRecommendation, setShowFirstRecommendation] = useState(true)
+    const [listActive, setListActive] = useState(false)
+    const [optionPointerIndex, setOptionPointerIndex] = useState(-1)
+    const [scrollWidth, setScrollWidth] = useState(null)
+    const [highlightedParts, setHighlightedParts] = useState([])
+    const [filteredRecommendations, setFilteredRecommendations] = useState(
```

Comment:
Was man hier noch machen könnte, wäre das wir die Recommendations nur generieren, wenn sich auch die die Eingabe geändert hat. Sowas erreicht man über useMemo:
```
   const filteredRecommendations = useMemo(() => {
       // hier kommt die Filterung von weiter unten
   }, [inputValue])
```
In dem Array gibt man, wovon die Berechnung abhängt. D.h. solange beim nächsten Rendern der gleiche inputValue wie beim letzten mal gilt, wird der zuletzt berechnete Wert genommen und keine neue Berechnung angestossen. Du brauchst dann auch die state-Variable nicht mehr